src/AWS.Deploy.ServerMode.Client/ServerModeSession.cs

```diff
@@ -0,0 +1,122 @@
+{
+    /// <summary>
+    /// Helper class that allows launching deployment tool in server mode.
+    /// </summary>
+    public class ServerModeSession
+    {
+        private const int TCP_PORT_ERROR = -100;
+
+        private readonly int _startPort;
+        private readonly int _endPort;
+
+        public ServerModeSession(int startPort = 5000, int endPort = 5100)
+        {
+            _startPort = startPort;
+            _endPort = endPort;
+        }
+
+        /// <summary>
+        /// Start deployment tool in server mode.
+        /// </summary>
+        /// <returns>Base URL of the  deployment tool local server.</returns>
+        /// <exception cref="InternalServerModeException">Throws when deployment tool server failed to start for un unknown reason.</exception>
+        /// <exception cref="PortUnavailableException">Throws when deployment tool server failed to start due to unavailability of free ports.</exception>
+        public async Task<string> Start()
+        {
+            var currentProcessId = Process.GetCurrentProcess().Id;
+
+            for (var port = _startPort; port <= _endPort; port++)
+            {
+                var baseUrl = $"http://localhost:{port}";
+
+                var command = $"dotnet aws server-mode --port {port} --parent-pid {currentProcessId}";
+                var startServerTask = RunCommand(command);
+
+                var isServerAvailableTask = IsServerAvailable(baseUrl);
+
+                // If the server is server health check succeeded
+                // return the baseUrl
+                if (isServerAvailableTask == await Task.WhenAny(startServerTask, isServerAvailableTask))
+                {
+                    return baseUrl;
+                }
+
+                // For -100 errors, we want to check all the ports in the configured port range
+                // If the error code other than -100, this is an unexpected exit code.
+                if (startServerTask.Result != TCP_PORT_ERROR)
+                {
+                    throw new InternalServerModeException($"\"{command}\" failed for unknown reason.");
+                }
+            }
+
+            throw new PortUnavailableException($"Free port unavailable in {_startPort}-{_endPort} range.");
+        }
+
+        private Task<bool> IsServerAvailable(string baseUrl)
+        {
+            var healthUrl = $"{baseUrl}/api/v1/health";
+            return WaitUntilHelper.WaitUntilSuccessStatusCode(healthUrl, TimeSpan.FromMilliseconds(100), TimeSpan.FromMinutes(10));
+        }
+
+        private async Task<int> RunCommand(string command)
+        {
+            var arguments = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? $"/c {command}" : $"-c \"{command}\"";
+
+            var processStartInfo = new ProcessStartInfo
+            {
+                FileName = GetSystemShell(),
+                Arguments = arguments,
+                UseShellExecute = false, // we don't want to start terminal
```

Comment:
We will want this configurable. So if we are in a diagnostic mode we can see the shell window.

Reply:
Added a flag which can be set from outside.